pkg/coscheduling/coscheduling.go

```diff
@@ -95,295 +80,316 @@ func New(obj runtime.Object, handle framework.FrameworkHandle) (framework.Plugin
-	minAvailable, exist := pod.Labels[PodGroupMinAvailable]
-	if !exist || len(minAvailable) == 0 {
-		return "", 0, nil
+// StartBatchSchedule receives the group name of pod to start scheduling
+func (cs *Coscheduling) StartBatchSchedule(fullName string) {
+	pgStatus := cs.pgMgr.PodGroupStatusCache().Get(fullName)
+	if pgStatus.PodGroup.Status.Phase != v1alpha1.PodGroupPreScheduling && pgStatus.PodGroup.Status.Phase != v1alpha1.
+		PodGroupScheduling {
+		return
 	}
-	minNum, err := strconv.Atoi(minAvailable)
+
+	// record time to avoid abnormal exit when bind
+	pg := pgStatus.PodGroup
+	pgCopy := pgStatus.PodGroup.DeepCopy()
+	pgCopy.Status.ScheduleStartTime = metav1.Now()
+
+	patch, err := util.CreateMergePatch(pg, pgCopy)
 	if err != nil {
-		klog.Errorf("PodGroup %v/%v : PodGroupMinAvailable %v is invalid", pod.Namespace, pod.Name, minAvailable)
-		return "", 0, err
+		return
 	}
-	if minNum < 1 {
-		klog.Errorf("PodGroup %v/%v : PodGroupMinAvailable %v is less than 1", pod.Namespace, pod.Name, minAvailable)
-		return "", 0, err
+
+	if err := cs.pgMgr.PatchPodGroup(pg.Name, pg.Namespace, patch); err != nil {
+		return
 	}
-	return podGroupName, minNum, nil
-}
 
-func (cs *Coscheduling) calculateTotalPods(podGroupName, namespace string) int {
-	// TODO get the total pods from the scheduler cache and queue instead of the hack manner.
-	selector := labels.Set{PodGroupName: podGroupName}.AsSelector()
-	pods, err := cs.podLister.Pods(namespace).List(selector)
-	if err != nil {
-		klog.Error(err)
-		return 0
+	// Start coscheduling
+	klog.V(4).Infof("Start scheduling %v", fullName)
+	localWaitingPods := cs.pgMgr.GetPodNodePairs(fullName)
+	if localWaitingPods == nil {
+		klog.V(4).Infof("Cannot found pending pods for %v", fullName)
+		return
+	}
+	localWaitingPodsCopy := localWaitingPods.Items()
+	if int32(len(localWaitingPodsCopy)) < pgStatus.PodGroup.Spec.MinMember-pgStatus.PodGroup.Status.Scheduled {
+		return
 	}
-	return len(pods)
+
+	// firstPodUID just record the first pod in the podsMap. When we delete any one of the pending pods,
+	// it would trigger evict of all cache.
+	firstPodUID := ""
+	for uid, pair := range localWaitingPodsCopy {
+		waitingPod := cs.frameworkHandler.GetWaitingPod(types.UID(uid))
+		// if in-tree WaitingPod timeout, we may get nil
+		if waitingPod == nil {
+			klog.V(4).Infof("Remove pod uid %v, pair %v, because waitingPod is nil", uid, pair.Object)
+			localWaitingPods.Delete(uid)
+			return
+		}
+		if firstPodUID == "" {
+			firstPodUID = uid
+		}
+		waitingPod.Allow(Name)
+		klog.V(4).Infof("Allowed pod %+v", pair.Object)
+	}
+	localWaitingPods.Delete(firstPodUID)
 }
 
-func (cs *Coscheduling) calculateBoundPods(podGroupName, namespace string) int {
-	nodeInfos, err := cs.frameworkHandle.SnapshotSharedLister().NodeInfos().List()
-	if err != nil {
-		klog.Errorf("Cannot get nodeInfos from frameworkHandle: %v", err)
-		return 0
+// rejectPod rejects pod in cache
+func (cs *Coscheduling) rejectPod(uid types.UID) {
```

Comment:
waitingPod reject will check if it is nil.

Reply:
You can not ensure  the `waitingPod reject` is always same. Framework interface is still alpha, so we should make the plugin as robust as possible. So I would like keep this

Reply:
It's amazing to me. This code base is release 1.19. You mean the interface will change in 1.19?

Reply:
No! You can take a look at the difference from 1.15 to 1.19.